File: orion-frontend/src/api/tenant.ts
/**
 * Tenant Management API Service
 * Multi-tenancy, quota management, and namespace pool
 */
import { api } from './client';

export interface TenantInfo {
  tenantId: number;
  userId?: string;
  roles?: string[];
  permissions?: string[];
}

export interface TenantQuota {
  tenantId: number;
  maxPipelines: number;
  maxPipelineRunsPerDay: number;
  maxConcurrentRuns: number;
  maxTasksPerPipeline: number;
  maxRunners: number;
  maxCpuCores: number;
  maxMemoryGb: number;
  maxStorageGb: number;
  maxNamespaces: number;
  apiRateLimit: number;
  apiRateLimitWindowSeconds: number;
}

export interface QuotaCheckResult {
  allowed: boolean;
  currentUsage: number;
  quotaLimit: number;
  remaining: number;
  message?: string;
}

export interface NamespacePoolEntry {
  id: string;
  namespaceName: string;
  clusterId: string;
  tenantId: number | null;
  status: 'available' | 'allocated' | 'reserved';
  purpose?: string;
  labels: Record<string, string>;
  allocatedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface NamespaceAllocationResult {
  success: boolean;
  namespace?: NamespacePoolEntry;
  error?: string;
}

export interface PoolStatus {
  totalNamespaces: number;
  availableNamespaces: number;
  allocatedNamespaces: number;
  reservedNamespaces: number;
  utilizationPercent: number;
}

// ==================== Tenant Context ====================

export function getTenantContext() {
  return api.get<{ context: TenantInfo | null }>('/api/tenant/context');
}

// ==================== Tenant Quota ====================

export function getTenantQuota(tenantId?: number) {
  return api.get<TenantQuota>('/api/tenant/quota', {
    headers: tenantId ? { 'x-tenant-id': tenantId.toString() } : {},
  });
}

export function updateTenantQuota(quota: Partial<TenantQuota>, tenantId?: number) {
  return api.put<TenantQuota>('/api/tenant/quota', quota, {
    headers: tenantId ? { 'x-tenant-id': tenantId.toString() } : {},
  });
}

export function checkTenantQuota(resourceType: string, amount: number, tenantId?: number) {
  return api.post<QuotaCheckResult>(
    '/api/tenant/quota/check',
    { resourceType, amount },
    {
      headers: tenantId ? { 'x-tenant-id': tenantId.toString() } : {},
    }
  );
}

// ==================== Namespace Pool ====================

export function getNamespacePoolStatus() {
  return api.get<PoolStatus>('/api/tenant/namespace/pool');
}

export function allocateNamespace(tenantId: string, namespaceType?: 'build' | 'deploy' | 'test') {
  return api.post<NamespaceAllocationResult>('/api/tenant/namespace/allocate', {
    tenantId,
    namespaceType,
  });
}

export function releaseNamespace(namespaceName: string) {
  return api.post<{ released: boolean }>('/api/tenant/namespace/release', { namespaceName });
}

export function getTenantNamespaces(tenantId: string) {
  return api.get<{ namespaces: NamespacePoolEntry[]; count: number }>(
    `/api/tenant/namespace/${tenantId}`
  );
}

// ==================== Middleware Config ====================

export function getMiddlewareConfig() {
  return api.get<{ config: { enabled: boolean; headerName: string; jwtTenantClaim: string } }>(
    '/api/tenant/middleware/config'
  );
}

export function updateMiddlewareConfig(config: {
  enabled?: boolean;
  headerName?: string;
  jwtTenantClaim?: string;
}) {
  return api.put('/api/tenant/middleware/config', config);
}

// ==================== Tenant CRUD ====================

export interface TenantEntity {
  id: string;
  name: string;
  display_name: string | null;
  status: string;
  settings: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface CreateTenantRequest {
  name: string;
  display_name?: string;
  settings?: Record<string, unknown>;
  autoAllocateNamespace?: boolean;
  initialNamespaceCount?: number;
  customQuota?: {
    maxPipelines?: number;
    maxPipelineRunsPerDay?: number;
    maxConcurrentRuns?: number;
    maxRunners?: number;
    maxCpuCores?: number;
    maxMemoryGb?: number;
    maxStorageGb?: number;
    maxNamespaces?: number;
  };
}

export interface PaginatedTenants {
  data: TenantEntity[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

export function listTenants(page = 1, limit = 20, status?: string) {
  const params: Record<string, string> = { page: String(page), limit: String(limit) };
  if (status) params.status = status;
  return api.get<PaginatedTenants>('/api/tenant', { params });
}

export function getTenant(id: string) {
  return api.get<TenantEntity>(`/api/tenant/${id}`);
}

export function createTenant(input: CreateTenantRequest) {
  return api.post<TenantEntity & { allocatedNamespaces?: NamespacePoolEntry[]; message?: string }>('/api/tenant', input);
}

export function updateTenant(id: string, input: Partial<CreateTenantRequest>) {
  return api.put<TenantEntity>(`/api/tenant/${id}`, input);
}

export function deleteTenant(id: string) {
  return api.delete(`/api/tenant/${id}`);
}

// ==================== Usage Statistics ====================

export interface ResourceUsage {
  used: number;
  limit: number;
}

export interface TenantUsage {
  usage: {
    pipelines: ResourceUsage;
    runners: ResourceUsage;
    namespaces: ResourceUsage;
    concurrentRuns: ResourceUsage;
    cpuCores: ResourceUsage;
    memoryGb: ResourceUsage;
    storageGb: ResourceUsage;
    pipelineRunsPerDay: ResourceUsage;
  };
  quota: TenantQuota;
}

export interface NamespaceUsageDetail {
  id: string;
  namespaceName: string;
  status: 'available' | 'allocated' | 'reserved';
  allocatedAt?: string;
  runnerCount: number;
  pipelineCount: number;
  activeRuns: number;
  cpuUsed: number;
  memoryUsed: number;
}

export function getTenantUsage(tenantId?: number) {
  return api.get<TenantUsage>('/api/tenant/usage', {
    headers: tenantId ? { 'x-tenant-id': tenantId.toString() } : {},
  });
}

export function getNamespaceUsageDetail(tenantId: string) {
  return api.get<{ namespaces: NamespaceUsageDetail[]; total: number }>(
    `/api/tenant/namespace/${tenantId}/usage`
  );
}

// ==================== Statistics ====================

export function getTenantStats(tenantId?: number) {
  return api.get<{
    stats: {
      tenantId: number;
      quotaUsage: Record<string, unknown>;
      namespaceCount: number;
    };
  }>('/api/tenant/count', {
    params: tenantId ? { status: undefined } : {},
    headers: tenantId ? { 'x-tenant-id': tenantId.toString() } : {},
  });
}

// ==================== Tenant Invites ====================

export interface TenantInvite {
  id: string;
  tenant_id: string;
  email: string;
  role: string;
  invite_code: string;
  status: 'pending' | 'accepted' | 'expired' | 'cancelled';
  invited_by: string;
  accepted_by?: string;
  expires_at: string;
  created_at: string;
  accepted_at?: string;
}

export interface CreateInviteRequest {
  email: string;
  role: string;
  message?: string;
}

export function inviteUser(tenantId: string, data: CreateInviteRequest) {
  return api.post<TenantInvite>(`/api/tenant/${tenantId}/invite`, data);
}

export function acceptInvite(code: string) {
  return api.post<{ success: boolean; tenant: TenantEntity; role: string; message: string }>(
    `/api/tenant/invite/${code}/accept`
  );
}

// ==================== Tenant Users ====================

export interface TenantUser {
  id: string;
  user_id: string;
  username: string;
  email: string;
  name?: string;
  role: string;
  status: string;
  last_login_at?: string;
  created_at: string;
}

export function getUsersByTenant(tenantId: string) {
  return api.get<{ users: TenantUser[]; total: number }>(`/api/tenant/${tenantId}/users`);
}

export function removeUserFromTenant(tenantId: string, userId: string) {
  return api.delete(`/api/tenant/${tenantId}/users/${userId}`);
}

// ==================== Tenant Alerts ====================

export interface TenantAlert {
  id: string;
  tenant_id: string;
  resource_type: string;
  threshold_percent: number;
  current_usage: number;
  quota_limit: number;
  notify_status: string;
  cooldown_until?: string;
  created_at: string;
}

export function getTenantAlerts(tenantId?: string, params?: { page?: number; limit?: number; resourceType?: string; status?: string }) {
  return api.get<{ alerts: TenantAlert[]; total: number; page: number; limit: number }>('/api/tenant/alerts', {
    params,
    headers: tenantId ? { 'x-tenant-id': tenantId } : {},
  });
}

export function getAlertStats(tenantId?: string) {
  return api.get<{
    stats: {
      byStatus: Record<string, number>;
      byResourceType: Record<string, number>;
      activeAlerts: TenantAlert[];
    };
  }>('/api/tenant/alerts/stats', {
    headers: tenantId ? { 'x-tenant-id': tenantId } : {},
  });
}

// ==================== Current Tenant ====================

export function getCurrentTenant() {
  return api.get<{
    tenant: TenantEntity;
    quota: TenantQuota;
    namespaceCount: number;
    namespaceLimit: number;
    activeAlertCount: number;
  }>('/api/tenant/current');
}

File: orion-tenant-svc-go/internal/models/tenant.go
package models

import "time"

// Tenant represents a multi-tenant account in Orion.
type Tenant struct {
	ID          string            `db:"id" json:"id"`
	Name        string            `db:"name" json:"name"`
	DisplayName *string           `db:"display_name" json:"display_name"`
	Status      string            `db:"status" json:"status"`
	Settings    map[string]any    `db:"settings" json:"settings"`
	CreatedAt   time.Time         `db:"created_at" json:"created_at"`
	UpdatedAt   time.Time         `db:"updated_at" json:"updated_at"`
}

// TenantUser represents a user-tenant membership relationship.
type TenantUser struct {
	ID              string    `db:"id" json:"id"`
	TenantID        string    `db:"tenant_id" json:"tenant_id"`
	UserID          string    `db:"user_id" json:"user_id"`
	Role            string    `db:"role" json:"role"`
	CreatedAt       time.Time `db:"created_at" json:"created_at"`
	UpdatedAt       time.Time `db:"updated_at" json:"updated_at"`
	// Populated via LEFT JOIN with users table
	Username        *string `db:"username" json:"username"`
	Email           *string `db:"email" json:"email"`
	UserDisplayName *string `db:"display_name" json:"display_name"`
	UserStatus      *string `db:"user_status" json:"user_status"`
}

// TenantInvite represents a pending user invitation to a tenant.
type TenantInvite struct {
	ID                string     `db:"id" json:"id"`
	TenantID          string     `db:"tenant_id" json:"tenant_id"`
	Email             string     `db:"email" json:"email"`
	Role              string     `db:"role" json:"role"`
	InviteCode        string     `db:"invite_code" json:"invite_code"`
	Status            string     `db:"status" json:"status"`
	InvitedBy         *string    `db:"invited_by" json:"invited_by"`
	AcceptedBy        *string    `db:"accepted_by" json:"accepted_by"`
	ExpiresAt         time.Time  `db:"expires_at" json:"expires_at"`
	CreatedAt         time.Time  `db:"created_at" json:"created_at"`
	AcceptedAt        *time.Time `db:"accepted_at" json:"accepted_at,omitempty"`
	// Populated via JOIN with tenants table
	TenantName        *string `db:"tenant_name" json:"tenant_name"`
	TenantDisplayName *string `db:"tenant_display_name" json:"tenant_display_name"`
}

// QuotaAlert represents a quota usage alert for a tenant.
type QuotaAlert struct {
	ID               string     `db:"id" json:"id"`
	TenantID         string     `db:"tenant_id" json:"tenant_id"`
	ResourceType     string     `db:"resource_type" json:"resource_type"`
	ThresholdPercent float64    `db:"threshold_percent" json:"threshold_percent"`
	CurrentUsage     float64    `db:"current_usage" json:"current_usage"`
	QuotaLimit       float64    `db:"quota_limit" json:"quota_limit"`
	NotifyStatus     string     `db:"notify_status" json:"notify_status"`
	CooldownUntil    *time.Time `db:"cooldown_until" json:"cooldown_until,omitempty"`
	CreatedAt        time.Time  `db:"created_at" json:"created_at"`
}

// TenantNamespace represents a K8s namespace allocation for a tenant.
type TenantNamespace struct {
	ID            string            `db:"id" json:"id"`
	NamespaceName string            `db:"namespace_name" json:"namespace_name"`
	ClusterID     string            `db:"cluster_id" json:"cluster_id"`
	TenantID      *string           `db:"tenant_id" json:"tenant_id"`
	Status        string            `db:"status" json:"status"` // available|allocated|reserved
	Purpose       *string           `db:"purpose" json:"purpose"`
	Labels        map[string]string `db:"labels" json:"labels"`
	AllocatedAt   *time.Time        `db:"allocated_at" json:"allocated_at,omitempty"`
	CreatedAt     time.Time         `db:"created_at" json:"created_at"`
	UpdatedAt     time.Time         `db:"updated_at" json:"updated_at"`
}

// QuotaConfig holds quota limits for a tenant.
type QuotaConfig struct {
	ID                        string            `db:"id" json:"id"`
	TenantID                  string            `db:"tenant_id" json:"tenant_id"`
	MaxPipelines              int               `db:"max_pipelines" json:"max_pipelines"`
	MaxPipelineRunsPerDay     int               `db:"max_pipeline_runs_per_day" json:"max_pipeline_runs_per_day"`
	MaxConcurrentBuilds       int               `db:"max_concurrent_builds" json:"max_concurrent_builds"`
	MaxTasksPerPipeline       int               `db:"max_tasks_per_pipeline" json:"max_tasks_per_pipeline"`
	MaxRunners                int               `db:"max_runners" json:"max_runners"`
	MaxCpuCores               int               `db:"max_cpu_cores" json:"max_cpu_cores"`
	MaxMemoryGb               int               `db:"max_memory_gb" json:"max_memory_gb"`
	MaxStorageMb              int               `db:"max_storage_mb" json:"max_storage_mb"`
	MaxProjects               int               `db:"max_projects" json:"max_projects"`
	MaxUsers                  int               `db:"max_users" json:"max_users"`
	ApiRateLimit              int               `db:"api_rate_limit" json:"api_rate_limit"`
	ApiRateLimitWindowSeconds int               `db:"api_rate_limit_window_seconds" json:"api_rate_limit_window_seconds"`
	Usage                     map[string]any    `db:"usage" json:"usage"`
	CreatedAt                 time.Time         `db:"created_at" json:"created_at"`
	UpdatedAt                 time.Time         `db:"updated_at" json:"updated_at"`
}

// RLSPolicy represents a PostgreSQL Row Level Security policy entry.
type RLSPolicy struct {
	ID              string `db:"id" json:"id"`
	TableName       string `db:"table_name" json:"table_name"`
	PolicyName      string `db:"policy_name" json:"policy_name"`
	SessionVariable string `db:"session_variable" json:"session_variable"`
	Enabled         bool   `db:"enabled" json:"enabled"`
}
